pkg/controllers/workgenerator/override.go

```diff
@@ -0,0 +1,209 @@
+				Version: selector.Version,
+				Kind:    selector.Kind,
+				Name:    selector.Name,
+			}
+			croMap[key] = append(croMap[key], snapshot)
+		}
+	}
+	klog.V(2).InfoS("Fetched clusterResourceOverrideSnapshots", "resourceBinding", klog.KObj(resourceBinding), "numberOfResources", len(croMap))
+	return croMap, nil
+}
+
+func (r *Reconciler) fetchResourceOverrideSnapshots(ctx context.Context, resourceBinding *placementv1beta1.ClusterResourceBinding) (map[placementv1beta1.ResourceIdentifier][]*placementv1alpha1.ResourceOverrideSnapshot, error) {
+	roMap := make(map[placementv1beta1.ResourceIdentifier][]*placementv1alpha1.ResourceOverrideSnapshot)
+
+	// For now, we get the snapshots sequentially. We can optimize this by getting them in parallel, but we need to reorder
+	// the snapshot lists saved in the map.
+	for _, namespacedName := range resourceBinding.Spec.ResourceOverrideSnapshots {
+		snapshot := &placementv1alpha1.ResourceOverrideSnapshot{}
+		if err := r.Client.Get(ctx, types.NamespacedName{Name: namespacedName.Name, Namespace: namespacedName.Namespace}, snapshot); err != nil {
+			if errors.IsNotFound(err) {
+				// It could be caused by that the user updates the override too frequently and the snapshot has been replaced
+				// by the new one.
+				// TODO: support customized revision history limit
+				klog.ErrorS(err, "The resourceOverrideSnapshot is deleted", "resourceBinding", klog.KObj(resourceBinding), "resourceOverrideSnapshot", namespacedName)
+				return nil, controller.NewUserError(fmt.Errorf("resourceSnapshot %s is not found", namespacedName))
+			}
+			klog.ErrorS(err, "Failed to get the resourceOverrideSnapshot",
+				"resourceBinding", klog.KObj(resourceBinding), "resourceOverrideSnapshot", namespacedName)
+			return nil, controller.NewAPIServerError(true, err)
+		}
+		for _, selector := range snapshot.Spec.OverrideSpec.ResourceSelectors {
+			key := placementv1beta1.ResourceIdentifier{
+				Group:     selector.Group,
+				Version:   selector.Version,
+				Kind:      selector.Kind,
+				Name:      selector.Name,
+				Namespace: snapshot.Namespace,
+			}
+			roMap[key] = append(roMap[key], snapshot)
+		}
+	}
+	klog.V(2).InfoS("Fetched resourceOverrideSnapshots", "resourceBinding", klog.KObj(resourceBinding), "numberOfResources", len(roMap))
+	return roMap, nil
+}
+
+func (r *Reconciler) applyOverrides(resource *placementv1beta1.ResourceContent, cluster clusterv1beta1.MemberCluster, croMap map[placementv1beta1.ResourceIdentifier][]*placementv1alpha1.ClusterResourceOverrideSnapshot, roMap map[placementv1beta1.ResourceIdentifier][]*placementv1alpha1.ResourceOverrideSnapshot) error {
+	var uResource unstructured.Unstructured
+	if err := uResource.UnmarshalJSON(resource.Raw); err != nil {
+		klog.ErrorS(err, "Work has invalid content", "selectedResource", resource.Raw)
+		return controller.NewUnexpectedBehaviorError(err)
+	}
+	gvk := uResource.GetObjectKind().GroupVersionKind()
+	key := placementv1beta1.ResourceIdentifier{
```

Comment:
this is not used if it's clusterScoped?

Reply:
it's used in line 122 when applying the clusterResourceOverrideSnapshots.